### ISSUE TITLE
ScrollView: Unsupported top level event "topScroll" dispatched

### ISSUE THREAD

**Issue by znacloud** [facebook/react-native#16522]
### Is this a bug report?
yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

yes

### Environment

  Environment:
    OS:  macOS Sierra 10.12.6
    Node:  8.4.0
    npm:  5.3.0
    Watchman:  4.7.0
    Xcode:  Xcode 8.3.3 Build version 8E3004b
    Android Studio:  2.3 AI-162.4069837

  Packages: (wanted => installed)
    react-native: 0.50.0.rc1
    react: 16.0.0

  Target Platform: Android (8.0.0)

  
### Steps to Reproduce
1. launch a scrollview page which has two Image as child
2. scroll this page
3.show red box: Unsupported top level event "topScroll" dispatched

**Comment by hramos** [comment]
Can you reproduce the issue using [Snack](http://snack.expo.io)? This step is necessary for people to be able to see and debug the issue being reported.

**Comment by znacloud** [comment]
Sorry, the target platform should be Android(8.0.8)

**Comment by znacloud** [comment]
@hramos Sorry, I'm not familiar with Snack。Here is exception stack:
`Unsupported top level event type "topScroll" dispatched
extractEvents
    ReactNativeFiber-dev.js:3519:22
extractEvents
    ReactNativeFiber-dev.js:3298:71
handleTopLevel
    ReactNativeFiber-dev.js:3539:64
<unknown>
    ReactNativeFiber-dev.js:3560:55
batchedUpdates
    ReactNativeFiber-dev.js:2754:26
batchedUpdates
    ReactNativeFiber-dev.js:202:35
batchedUpdatesWithControlledComponents
    ReactNativeFiber-dev.js:209:34
_receiveRootNodeIDEvent
    ReactNativeFiber-dev.js:3559:50
receiveEvent
    ReactNativeFiber-dev.js:3564:60
__callFunction
    MessageQueue.js:302:47
<unknown>
    MessageQueue.js:116:26
__guard
    MessageQueue.js:265:6
callFunctionReturnFlushedQueue
    MessageQueue.js:115:17
`

**Comment by ifsnow** [comment]
@hramos I am having the same problem. **I think horizontal scrollview has a trouble in all cases.**
It works well on react-native@0.49.5, but this error has occurred since version react-native@0.50.0-rc.0

The following simple code also causes an error when scrolling. tested on API 23(Android 6.0) and 24(Android 7.0).

```js
import React, { Component } from 'react';
import {
  View,
  StyleSheet,
  ScrollView
} from 'react-native';

export default class ScrollViewTest extends Component {
  render() {
    return (
      <View style={styles.container}>
        <ScrollView horizontal>
          <View style={styles.dummy}></View>
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  dummy: {
    flex: 1,
    width: 1000,
    height: 400,
    backgroundColor: 'red'
  }
});
```

**Comment by codeYDH** [comment]
@hramos I am having the same problem.

**Comment by rabingaire** [comment]
I also have the same issue !!! Please fix this fast.

**Comment by bvaughn** [comment]
I believe this is related to changes on the native (`UIManager`) side and should be resolved with commit fdf0495.

**Comment by Kottidev** [comment]
@hramos we need to fix it quickly , many apps use ScrollView  with horizontal=true

**Comment by ide** [comment]
Cherry-picked https://github.com/facebook/react-native/commit/1b71e03932f44e212b297b2c1e02100b6de74b93 into 0.50.2, which is now on npm.

**Comment by amirkheirabadi** [comment]
I still have this problem into 0.50.2

**Comment by awnali** [comment]
i also faced same issue and then changed the react-native version from 0.50 to 0.49.* and everything works.

**Comment by PetrVasiliev** [comment]
Yes! I also have this very bad problem, which killed my app

**Comment by kelset** [comment]
Updating to `0.50.3` fixed it for me - thanks for the quick fix!

**Comment by Omniwheel** [comment]
Still having the issue in 0.50.3.

**Comment by donnes** [comment]
Here too, upgrading to `0.50.3` don't solved the problem for me.

**Comment by rambabusaravanan** [comment]
I'm also facing the same error "_unsupported top level  event type "topScroll" dispatched_" when using `scrollEnabled: true` for TabNavigator in `react-navigation` which internally is a ScrollView

This error occurs in latest react-native@0.50.3 but works fine in react-native@0.49.0 (which is expo's latest supported version). Hence no issues in snack / can't show demo in [snack](http://snack.expo.io) @hramos

**Comment by booleanBoy** [comment]
0.50.3 hasn't solved the problem for me either. 

I'm also finding that while I can still work on the emulator (as long as I don't scroll) the signed APK I'm building crashes when run on a device. Has anyone else experience this or is it an unrelated problem?

**Comment by Rom96** [comment]
You can use ViewPagerAndroid instead of ScrollView but this works only with Android devices.

**Comment by iRoachie** [comment]
Seeing this bug on iOS as well on 0.50.3

**Comment by booleanBoy** [comment]
Thanks @Rom96, that's a great temporary workaround for me as I'm only working on Android 👍

**Comment by bvaughn** [comment]
Hi @booleanBoy / @rambabusaravanan / @donnes / @Omniwheel 

I'm confused by the fact that you're still seeing this issue and I'd like to get a little more background on what you're doing.

Taking `ScrollView` and the "topScroll " event type as an example. [`ScrollView` registers its native code](https://github.com/facebook/react-native/blob/master/Libraries/Components/ScrollView/ScrollView.js#L842-L876) using `requireNativeComponent` which [mixes in a set of shared, default event types](https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/requireNativeComponent.js#L64) which [in this case](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L260) [includes the "topScroll" event](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L260) (among others). Provided you're accessing `ScrollView` from the [`react-native-implementation`](https://github.com/facebook/react-native/blob/master/Libraries/react-native/react-native-implementation.js) then I think this should just work.

Are you by chance loading the view some other way? eg directly accessing `UIManager["ScrollView"]`?

**Comment by Amurmurmur** [comment]
Same issue.
```
Unsupported top level event type "topScrollBeginDrag" dispatched
ExceptionsManager.js:65
Unsupported top level event type "topScroll" dispatched
ExceptionsManager.js:65
Unsupported top level event type "topScroll" dispatched
ExceptionsManager.js:65
Unsupported top level event type "topScrollEndDrag" dispatched
ExceptionsManager.js:65
Unsupported top level event type "topMomentumScrollBegin" dispatched
ExceptionsManager.js:65
Unsupported top level event type "topScroll" dispatched
ExceptionsManager.js:65
Unsupported top level event type "topMomentumScrollEnd" dispatched
ExceptionsManager.js:65
```
<img width="482" alt="screen shot 2017-11-13 at 17 58 27" src="https://user-images.githubusercontent.com/8608333/32740874-49e6f72a-c89c-11e7-8848-890b579f53e5.png">

**Comment by bvaughn** [comment]
@Amurmurmur Please see my question above your comment.

**Comment by iCodePup** [comment]
Same issue using react-native 0.50.3 and
https://github.com/react-native-community/react-native-tab-view
With scrollEnabled true.
Problem does not appear with 0.49.3

**Comment by Amurmurmur** [comment]
@bvaughn I use FlatList the following way:
```
  render () {
    let items = this.props.children

    return (
      <FlatList
        style={styles.list}
        data={items}
        onMomentumScrollEnd={this.onScrollEnd}
        keyExtractor={(item) => items.indexOf(item)}
        pagingEnabled
        horizontal
        renderSeparator={() => null}
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={false}
        directionalLockEnabled
        renderItem={this._renderItem} />
    )
  }
```

**Comment by shawngustaw** [comment]
I noticed this when fooling around with react-navigation (w/ RN 50.0.3). Can reproduce by doing a simple Tab Navigator and doing a press and hold + drag gesture on a Tab. It's pretty easy to setup (I can paste code if needed) so hopefully that helps with reproducing.

**Comment by ifsnow** [comment]
@bvaughn As I mentioned before, the following simple code still has an error when scrolling. tested on API 23(Android 6.0) and 24(Android 7.0).

**I think horizontal scrollview has a trouble in all cases**

```js
import React, { Component } from 'react';
import {
  View,
  StyleSheet,
  ScrollView
} from 'react-native';

export default class ScrollViewTest extends Component {
  render() {
    return (
      <View style={styles.container}>
        <ScrollView horizontal>
          <View style={styles.dummy}></View>
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  dummy: {
    flex: 1,
    width: 1000,
    height: 400,
    backgroundColor: 'red'
  }
});
```

**Comment by mhemrg** [comment]
Same issue. I changed the react-native version in the `package.json` file and now it works fine:
```
"react-native": "^0.49.5"
```

**Comment by meurling** [comment]
I am getting this error when trying to drag up my bottom tabbar, which shouldn't result in anything happening at all!

**Comment by cleandersonlobo** [comment]
> I am getting this error when trying to drag up my bottom tabbar, which shouldn't result in anything happening at all!

Same issue.

` "react-native": "^0.50.3"`

**Comment by MedericPixium** [comment]
Same issue on 
`"react-native": "^0.50.3"`
using a horizontal Scrollview for my carousel

**Comment by Sergeypt423** [comment]
Any updates with this issue?

I really need it to be done asap.

**Comment by chrispaynter** [comment]
I'm also having this issue. There is a modal on the same page that is triggered by a button. If I open the modal and then close it, the `Unsupported top level event type "topScrollBeginDrag" dispatched` no longer appears and my caurosel works as expected.

**Comment by lee-d** [comment]
We as well experience the same error with v0.50.2. We implemented a pageable horizontal ScrollView. iOS-Simulator works fine. Only Android-Simulator breaks on interaction.

**Comment by MedericPixium** [comment]
@Sergeypt423 Agreed we are supposed to launch our app by the end of the month and it would be very nice to have it fixed by then T_T

**Comment by legion-zver** [comment]
Same issue. Use 0.50.3 version, and only on Android this problem (((

**Comment by nimamyscreen** [comment]
same issue

**Comment by pangpond** [comment]
me too with
"react": "16.0.0",
"react-native": "0.50.3",

**Comment by cleandersonlobo** [comment]
temporary solution:
I changed the react-native version in the package.json file and now it works fine:
```
"react-native": "0.49.5",
```

**Comment by ddman** [comment]
me too with
		"react": "16.0.0",
		"react-native": "0.50.1",

**Comment by ide** [comment]
Just to set expectations, I am not aware of anyone looking into this issue. Having been in similar situations in the past, two things that have worked well to get unblocked are forking+fixing the issue locally and contracting an expert.

**Comment by rolignu2** [comment]
same issue 

"react": "16.0.0",
"react-native": "0.50.3"

**Comment by bvaughn** [comment]
> Just to set expectations, I am not aware of anyone looking into this issue.

This is mostly true. Dmitry and I have talked about it a few times but neither of us have had time to look into it yet. It's on my wish list for things I'd like to find time to do but I don't know when that will happen yet.

If someone in the community has bandwidth to look into it - that would be ideal.

**Comment by ifsnow** [comment]
@bvaughn @ide 

I have found the cause of error in react-native@0.50.3 and how to fix it. However, I realized that there was no problem in current master branch, so I think it doesn't make sense to solve the issue in 0.50.3 branch.

# current status
- ReactNativeRenderer-dev.js : no problem (master)
- ReactNativeFiber-dev.js : error (v0.50.3)

# the cause of error
Horizontal scrollview uses `AndroidHorizontalScrollView` instead of `RCTScrollView`. The problem is that `AndroidHorizontalScrollView` does not register custom direct events (topScroll, topScrollBeginDrag, ...). 

The events should be registered during the initialization of the three modules(RCTScrollView, 
 AndroidHorizontalScrollView, AndroidHorizontalScrollContentView), but it seems to be a problem.

Adding the following code to `ReactHorizontalScrollViewManager` solves the problem.

```java
@Override
public @Nullable Map getExportedCustomDirectEventTypeConstants() {
  return createExportedCustomDirectEventTypeConstants();
}

public static Map createExportedCustomDirectEventTypeConstants() {
  return MapBuilder.builder()
    .put(ScrollEventType.SCROLL.getJSEventName(), MapBuilder.of("registrationName", "onScroll"))
    .put(ScrollEventType.BEGIN_DRAG.getJSEventName(), MapBuilder.of("registrationName", "onScrollBeginDrag"))
    .put(ScrollEventType.END_DRAG.getJSEventName(), MapBuilder.of("registrationName", "onScrollEndDrag"))
    .put(ScrollEventType.MOMENTUM_BEGIN.getJSEventName(), MapBuilder.of("registrationName", "onMomentumScrollBegin"))
    .put(ScrollEventType.MOMENTUM_END.getJSEventName(), MapBuilder.of("registrationName", "onMomentumScrollEnd"))
    .build();
}
```

I think it is necessary to improve the initialization part of `ScrollView`. I will work on this later.

I hope this helps.

**Comment by MedericPixium** [comment]
@ifsnow I have modified the file in my node module however the error remains. do I have to recompile from the src?

**Comment by ifsnow** [comment]
@MedericPixium Check out [this manual](https://facebook.github.io/react-native/docs/android-building-from-source.html) first.

**Comment by bvaughn** [comment]
Thanks for digging into this issue and sharing your findings, @ifsnow ❤️ 

I'm still concerned that it addresses a symptom without understanding the underlying cause though. As I mentioned in a previous comment, events like "_topScroll_", "_topScrollBeginDrag_", etc. are actually _supposed_ to be [available for all view manager types by default](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L90) because of [how this code works](https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/requireNativeComponent.js#L64).

Maybe I'm misunderstanding your comment?

**Comment by bvaughn** [comment]
I just ran the open-source [`RNTester`](https://github.com/facebook/react-native/tree/master/RNTester) app on Android but was unable to reproduce the error being reported. I ran it from both `master` and the [`0.50-stable` branch](https://github.com/facebook/react-native/tree/0.50-stable) (where 0.53 was released from). Adding this to the fact that internally (at Facebook) we aren't seeing this problem- and I'm not sure of how best the proceed.

**Comment by K-Leon** [comment]
Well i think there’s really something off. I‘m seeing this one very often:

Unsupported top level event type "topTouchStart" dispatched Unsupported top level event type "topTouchStart" dispatched
 
I believe it‘s not so specific to ScrollView, but more related to a behavioral change of the validation / Check Code (maybe with React 16 or an assert that didn“t fire before?)

**Comment by ifsnow** [comment]
@bvaughn I also wonder why suddenly there is a problem. Unfortunately, at least in my environment, errors still occur. I looked a little bit further and hope it helps.

# Below is the process I understood.

1. Creates native component using `requireNativeComponent`, then `createReactNativeComponentClass` is called internally.
https://github.com/facebook/react-native/blob/v0.50.3/Libraries/ReactNative/requireNativeComponent.js#L196

2. View’s config is saved in internal map to load lazy.
https://github.com/facebook/react-native/blob/v0.50.3/Libraries/Renderer/ReactNativeFiber-dev.js#L2966

3. View is created and is registered events when creating instance. (createInstance)
https://github.com/facebook/react-native/blob/v0.50.3/Libraries/Renderer/ReactNativeFiber-dev.js#L3054

# Take a closer look

- Sample code that causes the problem.
```js
render() {
  return (
    <ScrollView horizontal>
      <View style={styles.dummy}></View>
    </ScrollView>
  );
}
```

- The process of creating a view
1. createReactNativeComponentClass : RCView
2. createReactNativeComponentClass : RCTImageView
3. createReactNativeComponentClass : RCTTextInlineImage
4. createReactNativeComponentClass : RCTScrollView
5. createReactNativeComponentClass : AndroidHorizontalScrollView
6. createReactNativeComponentClass : AndroidHorizontalScrollContentView
7. createInstance : RCView
7-1. call getViewConfig() of createReactNativeComponentClass
7-2. **try to register default event types, but UIManager.ViewManagerNames is undefined, so the default events are not registered.**
https://github.com/facebook/react-native/blob/v0.50.3/Libraries/ReactNative/requireNativeComponent.js#L62
8. createInstance : AndroidHorizontalScrollContentView
9. createInstance : AndroidHorizontalScrollView

`RCTScrollView` is not called because it is not created. It has override method to register events(topScroll,...).

Why `UIManager.ViewManagerNames` is undefined seems to be a clue to solving.

**Comment by bvaughn** [comment]
> 7-2. try to register default event types, but UIManager.ViewManagerNames is undefined, so the default events are not registered.

Ooh! Thanks @ifsnow! That is actually useful info. Let me chat with Dmitry about it.

**Comment by comfly** [comment]
The default event types should be registered in case UIManager.ViewManagerNames is undefined. The `else` clause there merges in events from a ViewManager and UIManager.genericBubblingEventTypes/genericDirectEventTypes which are default and exported with UIManagerModule creation.

**Comment by ifsnow** [comment]
@bvaughn @comfly That’s right. This is an important part. I think I found a clue to solving this mystery.

There are no 5 events(topScroll, …) in `getDirectEventTypeConstants` of 0.50.3 branch.

# 0.50.3 branch
https://github.com/facebook/react-native/blob/v0.50.3/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L83

# master branch
https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L78

**Comment by bvaughn** [comment]
Ooh snap! So the native code is trailing behind in this branch!

We added those events types _after_ (ca834f9 when we realized that several viewmanagers didn't explicitly define them.

**Comment by bvaughn** [comment]
cc @grabbou Can we get that change (ca834f9) cherry-picked into the 0.50 branch and released as 0.50.4?

**Comment by grabbou** [comment]
Going to do that right now.

**Comment by kelset** [comment]
Just tested `0.50.4` and the error is no more.

Thanks @bvaughn & @grabbou !

**Comment by asgvard** [comment]
Perfect, Thanks! Funny thing that I've just encountered this issue for the first time like 30 minutes ago, googled for this issue, and saw that it's going to be released pretty much now hahah :)

**Comment by mc-tracker** [comment]
could anyone please explain this case?In react-native 0.54
![image](https://user-images.githubusercontent.com/22979088/33226775-03ed7ea0-d1d0-11e7-9cae-1a8393c9c8a0.png)

**Comment by Amurmurmur** [comment]
@mc-tracker Reset cache.
Please run:
```
npm start -- --reset-cache
```

**Comment by mc-tracker** [comment]
@Amurmurmur rebuild apk resolved it , thanks anyways

**Comment by MedericPixium** [comment]
Thank you, updated to 0.54 and fixed it

**Comment by K-Leon** [comment]
@bvaughn Thank‘s for this fix! But it seems something is still off on a different place: i‘m seeing this issue with `topTouchStart`

**Comment by bvaughn** [comment]
@K-Leon "_topTouchStart_" ([`TouchEventType.START`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/TouchEventType.java#L16)) is specified as one of the default bubbling event types (see [`getBubblingEventTypeConstants`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L27-L40)). This means it should always be mixed in, by one of the following code paths:

Are views are being lazy-loaded?
* **Yes**: 
[`requireNativeComponent`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/Libraries/ReactNative/requireNativeComponent.js#L64) >
[`UIManager.getDefaultEventTypes`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L259-L260) >
[`UIManagerModuleConstantsHelper.getDefaultExportableEventTypes`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstantsHelper.java#L44-L46) >
[`UIManagerModuleConstants.getBubblingEventTypeConstants`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L27-L40).

* **No**:
[`requireNativeComponent`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/Libraries/ReactNative/requireNativeComponent.js#L66-L69) >
[`UIManager.genericBubblingEventTypes`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstantsHelper.java#L71-L108) >
[`UIManagerModuleConstants.getBubblingEventTypeConstants`](https://github.com/facebook/react-native/blob/554e873f585ea05ce1e9fe4ff71173c47b3c259c/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L27-L40)

Quickly spot-checking the `v0.50.4` branch, all of the above codepaths seem to be in-sync with `master`.

**Comment by bvaughn** [comment]
Did you try resetting cache (as specified [above](https://github.com/facebook/react-native/issues/16522#issuecomment-346939396))?
```
npm start -- --reset-cache
```

**Comment by meinto** [comment]
Yea version `v0.50.4` fixed the issue :+1: 

In my case to get it work, i had to remove to `android/app/build` folder first to remove already built apks.

**Comment by hssrrw** [comment]
I've just started a new project with React Native `0.50.4`. I see the error on iOS every time pressing Back button in toolbar or performing swipe-back gesture in `NavigatorIOS` component.

![photo_2017-12-05_12-38-23](https://user-images.githubusercontent.com/1917763/33603036-eb97c2ce-d9b9-11e7-97c9-0a14f253224b.jpg)

Can I fix this the way it fixed in Android? Can't find the place with UIManager constants in iOS sources though.

**Comment by bvaughn** [comment]
I'm not very familiar with the native iOS code. But it looks like [`RCTComponentData.m` gathers native events](https://github.com/facebook/react-native/blob/71006f74cdafdae7212c8a10603fb972c6ee338c/React/Views/RCTComponentData.m#L417-L425) and exposes them to [`RCTUIManager.m` which passes on to JS](https://github.com/facebook/react-native/blob/71006f74cdafdae7212c8a10603fb972c6ee338c/React/Modules/RCTUIManager.m#L1461-L1506). I added some logging to [`requireNativeComponent`](https://github.com/facebook/react-native/blob/71006f74cdafdae7212c8a10603fb972c6ee338c/Libraries/ReactNative/requireNativeComponent.js#L109) just to double check the event types it gets, and it shows that `RCTScrollView` specifies `topScroll` as one of its `directEventTypes`. Seems like it should work fine?

**Comment by Narutuffy** [comment]
If someone is facing the same problem in expo, try updating expo SDK to 23 and the expo app too. I didn't update the app and lost a lot of time on that.

**Comment by JoseVf** [comment]
Just created a new project with react-native init and get this error 
![screenshot_1513088918](https://user-images.githubusercontent.com/22102736/33889679-e01aaa5a-df1e-11e7-873c-7ec00eb7fbea.png)

react-native: 0.51.0
react: 16.0.0
device: Nexus 5 API 23 (AVD)

**Comment by K-Leon** [comment]
@bvaughn I'm still seeing it on android - event after cache clear. I can only reproduce it in production - i've still no idea where it is triggered. Even though this crash is new and currently the most "popular".

**Comment by nhoxbypass** [comment]
still happen in RN 0.51

**Comment by 4thjuly** [comment]
My app was working fine until I updated to Expo 2.2.0 on my phone and now I get this error

**Comment by marc9000** [comment]
Still happening for 0.51.0. Confirming that moving to ^0.49.5 does the trick. 

It seems like it's for all sorts of components, e.g. happened to TouchableOpacity for me.

**Comment by dcu-sharepoint** [comment]
@ifsnow; now that you can @override whatever function you're overriding, what do the rest of us mere mortals do to fix our issue the with the red horror screen from an app update. In my case, all I did was add a StatusBar component without any code yet!.... go figure, add one problem, get may problems back.... as usual, an app update that shows up as an error for the rest of us.....

**Comment by ifsnow** [comment]
@marc9000 @dcu-sharepoint @nhoxbypass I don't get any more errors in RB 0.51. 
Could you show me the code that caused the problem? or Try the code below in your new project.

In my case, It works well in RB 0.51(Android 6.0)
```js
import React, { Component } from 'react';
import {
  View,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
  Text,
  StatusBar
} from 'react-native';

export default class App extends Component {
  render() {
    return (
      <View style={styles.container}>
        <StatusBar
          backgroundColor="blue"
          barStyle="light-content"
        />

        <TouchableOpacity>
          <Text>Click</Text>
        </TouchableOpacity>

        <ScrollView horizontal>
          <View style={styles.dummy}></View>
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  dummy: {
    flex: 1,
    width: 1000,
    height: 400,
    backgroundColor: 'red'
  }
});
```

**Comment by meurling** [comment]
Hi! I use RN --V0.51 and get the same error with the following code:

```
javascript
<View>
                <ScrollView 
                    horizontal={true}
                    style={{marginBottom: 3}}
                    showsHorizontalScrollIndicator={false}
                >
                    <View style={styles.circleMargin} >
                    <Text style={styles.categoryText}> Drinks </Text>
                        <TouchableOpacity onPress={() => setCategory("drinks", selectedCategory)} >
                            <View style={styles.circle} >
                                <Entypo 
                                    name="drink" 
                                    size={categorySize} 
                                    color={ (selectedCategory === "drinks") ? selectedColor : iconColor} 
                                    style={styles.categoryIcon} />
                            </View>
                        </TouchableOpacity>
                    </View>
                 </ScrollView>
            </View>
```

**Comment by cyphire** [comment]
Is my issue related to this?  I am in 0.51 and when I build an app for production I get a crash on startup (my first screen has a ScrollView with no parameters).

I traced the error to: Subscribable.js:

the this.subscribableSubscriptions is null, and it shouldn't be.  That is why the crash comes on the ForEach.

Here is the current code...
```
  componentWillUnmount: function() {
      this._subscribableSubscriptions.forEach(
      (subscription) => subscription.remove()
    );
     this._subscribableSubscriptions = null;
  },
```
When I modify the code to this, the crash goes away...
  
```
componentWillUnmount: function() {
  console.log(`Subscribable - error location? ${this._subscribableSubscriptions}`)  
  if (false) {
    this._subscribableSubscriptions.forEach(
      (subscription) => subscription.remove()
    );
  }
    this._subscribableSubscriptions = null;
  },
```
What do you think?

**Comment by avshenuk** [comment]
@cyphire The same problem here. Release app without any big changes from my side started to throw the same error:
```
ReactNative: TypeError: undefined is not an object (evaluating 

'this._subscribableSubscriptions.forEach')

This error is located at:
  in ScrollView
in t
in t
in RCTView
in r
in Connect(r)
in RCTView
in a
in n
in t
in RCTView
in RCTView
in t, stack:
componentWillUnmount@210:183
<unknown>@153:2884
<unknown>@41:51861
mountClassInstance@41:51916
<unknown>@41:58934
beginWork@41:58969
q@41:88677
ee@41:89278
ne@41:90352
ue@41:92806
ae@41:92107
enqueueSetState@41:25532
setState@48:846
onStateChange@303:5122
```

**Comment by avshenuk** [comment]
Decreased the version of React native to 0.49.5 and it works now.
@cyphire Do you think we should create an issue ?

**Comment by cyphire** [comment]
I was going to, almost had it finished, but wanted to explore the root causes of it.

Do you happen to use ignite?  I was wondering if it was some middleware or other issue.

It seems to me that there might be issues with anything which uses keyboard events (dismissing the keyboard) and ScrollView.  I moved my code to FlatList instead, but same error - I guess they are all based on ScrollView.  Did you also use  keyboard events to get this error?   The problem is definitely failing in the Scribbable.js module - but that's not the problem, it's whatever tries to create the events..

Here is what I was going to post:  If you want to update the below and post it that would be fine, either  way.



### Is this a bug report?

Yes



### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment
                                                       
Environment:                                           
  OS: Windows 10                                       
  Node: 8.9.1                                          
  Yarn: 1.3.2                                          
  npm: 5.5.1                                           
  Watchman: Not Found                                  
  Xcode: N/A                                           
  Android Studio: Version  3.0.0.0 AI-171.4408382      
                                                       
Packages: (wanted => installed)                        
  react: 16.0.0 => 16.0.0                              
  react-native: 0.51.0 => 0.51.0                       
                                                       

### Steps to Reproduce

1.  When building the app in normal debug mode, no issue. 
2.  In the "android" directory, I ran "gradlew assembleRelease".
3.  This created the correct SIGNED app-release.apk in the apk directory.
4.  Installed the app-release.apk in both the emulator on computer as well as an LG V20 device.
5.  Both the emulator and the device crash on startup.

### Expected Behavior



I expected that the release version would work as the debug version did.

### Actual Behavior

Here is where the error occurs: in Subscribable.js 
(..\node_modules\react-native\Libraries\Components\Subscribable.js)

This is the original code.
```
  componentWillUnmount: function() {
    this._subscribableSubscriptions.forEach(
      (subscription) => subscription.remove()
    );
    this._subscribableSubscriptions = null;
  },

```
Here is where i removed the failing call.  I have confirmed that this._subscribableSubscriptions returns a null and crashes when the forEach() is called.
```
  componentWillUnmount: function() {
  console.log(`Subscribable - error location? ${this._subscribableSubscriptions}`)  
  if (false) {
    this._subscribableSubscriptions.forEach(
      (subscription) => subscription.remove()
    );
  }
    this._subscribableSubscriptions = null;
  },

```
Following removing the offending code, the app no longer crashes on startup.

FROM LOGCAT:

11:01:04.208  5499  5499 D ReactNative: ReactInstanceManager.ctor()
12-25 11:01:04.212  5499  5499 D ReactNative: ReactInstanceManager.createReactContextInBackground()
12-25 11:01:04.212  5499  5499 D ReactNative: ReactInstanceManager.recreateReactContextInBackgroundInner()
12-25 11:01:04.213  5499  5499 D ReactNative: ReactInstanceManager.recreateReactContextInBackgroundFromBundleLoader()
12-25 11:01:04.213  5499  5499 D ReactNative: ReactInstanceManager.recreateReactContextInBackground()
12-25 11:01:04.213  5499  5499 D ReactNative: ReactInstanceManager.runCreateReactContextOnNewThread()
12-25 11:01:04.246  5499  5515 D ReactNative: ReactInstanceManager.createReactContext()
12-25 11:01:04.389  5499  5515 D ReactNative: Initializing React Xplat Bridge.
12-25 11:01:04.394  5499  5515 D ReactNative: Initializing React Xplat Bridge before initializeBridge
12-25 11:01:04.413  5499  5515 D ReactNative: Initializing React Xplat Bridge after initializeBridge
12-25 11:01:04.414  5499  5515 D ReactNative: CatalystInstanceImpl.runJSBundle()
12-25 11:01:04.454  5499  5524 D ReactNative: ReactInstanceManager.setupReactContext()
12-25 11:01:04.454  5499  5524 D ReactNative: CatalystInstanceImpl.initialize()
12-25 11:01:04.459  5499  5524 D ReactNative: ReactInstanceManager.attachRootViewToInstance()
12-25 11:01:04.927  5499  5522 I ReactNativeJS: Running application "BinStuffer" with appParams: {"rootTag":1}. __DEV__ === false, development-level warning are OFF, performance optimizations are ON
12-25 11:01:04.995  5499  5522 E ReactNativeJS: TypeError: undefined is not an object (evaluating 'this._subscribableSubscriptions.forEach')
12-25 11:01:04.995  5499  5522 E ReactNativeJS:
12-25 11:01:04.995  5499  5522 E ReactNativeJS: This error is located at:
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in ScrollView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in Connect(t)
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in n
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in n
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in n
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in r
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in Unknown
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in a
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in Connect(t)
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in Connect(t)
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in n
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:04.995  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS: TypeError: undefined is not an object (evaluating 'this._subscribableSubscriptions.forEach')
12-25 11:01:05.003  5499  5522 E ReactNativeJS:
12-25 11:01:05.003  5499  5522 E ReactNativeJS: This error is located at:
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in ScrollView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in Connect(t)
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in n
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in n
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in n
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in r
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in Unknown
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in a
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in Connect(t)
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in Connect(t)
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in n
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in RCTView
12-25 11:01:05.003  5499  5522 E ReactNativeJS:     in t
12-25 11:01:05.336  5499  5499 D ReactNative: ReactInstanceManager.detachViewFromInstance()                   

### Reproducible Demo



(Paste the link to an example project and exact instructions to reproduce the issue.)

**Comment by dody87** [comment]
I'm using NavigatorIOS and also same problem when goes back.

**Comment by cyphire** [comment]
I have created an issue:

https://github.com/facebook/react-native/issues/17358

**Comment by avshenuk** [comment]
Thanks Brian. Actually there is already an issue:
https://github.com/facebook/react-native/issues/17348

**Comment by cyphire** [comment]
Closed my issue for the other one.

_________________________________________
Brian B. Canin    brian.canin@gmail.com

858-945-5601 (cell)

**Comment by otoinsa** [comment]
I have the same issue, using NavigatorIOS and pressing back button/swiping back to previous view. On dev builds works fine, however on production builds it just outright crashes. In my case there are no scroll views involved, just some View's with 1 image between them.

I'm on 0.51 and would really like to get rid of this nasty bug,
come to think of it - how come no one noticed it and released another React Native version?
Each update makes React Native a little better and a lot worse :D FML, I use this in production ;(

**Comment by smhmayboudi** [comment]
I am a new bee worker and I follow these instruction:

1. create-react-native-app testApp
2. cd testApp
3. yarn upgrade
4. [I copy the last sample code at https://facebook.github.io/react-native/docs/navigation.html to app.js]
5. yarn start
6. start Expo app on iOS
7. scan the barcode
8. I got RDS (RED DEAD SCREEN) !?

**Comment by peterlevel1** [comment]
## a magic tour for this bug

also meet the bug which the same to the title: Unsupported top level event "topScroll" .
Now I am gonna tell you a tale about what I met today !

when I met the bug this afternoon, I do the search work: Unsupported top level event "topScroll" ,
and I came here to see you guys talking about this issue, I saw some guy said react-native@0.49.5 would fix this problem, so: yarn remove react-native --save && yarn add react-native@0.49.5 --save, and it just can not downloaded, so I changed the tool to be cnpm: cnpm i react-native@0.49.5, and it was downloaded. I met depressions when react-native run-ios building the package, so I run: cnpm remove react-native --save. I like react-native, so it can not stop my steps,
I planned to do some fallbacks that I just do not click the back button on the simulator until facebook engineers handle this bug. so I run: yarn add react-native --save(NOTICE: I do not enter the version, it would be the latest !). 

the magic things happen! it just work fine! the latest version of react-native is 0.51.0 !

### what I only want is just back to the prev page, the codes cause that the back button of the navigator can not be tapped, you would receive these codes if you do that:
#### Unsupported top level event "topScroll" .

```javascript: 
this.props.navigator.push({
    component: SecondScene,
    title: 'Second Scene !!!'
});
```

So, is it magic ?

**Comment by mjugger** [comment]
I updated to `0.50.4` and ran `npm start-- --reset-cache`, but I still get the `topScrollBeingDrag` issue.
I'm on android and my dependancies are as follows:
`  "dependencies": {
    "d3-array": "^1.2.1",
    "d3-scale": "^1.0.7",
    "d3-shape": "^1.2.0",
    "expo": "23.0.4",
    "intl": "^1.2.5",
    "native-base": "^2.3.6",
    "prop-types": "^15.6.0",
    "react": "16.0.0",
    "react-native": "0.50.4",
    "react-native-easy-grid": "^0.1.15",
    "react-native-slider": "^0.11.0",
    "react-native-vector-icons": "^4.4.2",
    "react-navigation": "^1.0.0-beta.21",
    "react-redux": "^5.0.6",
    "redux": "^3.7.2",
    "redux-saga": "^0.16.0"
  }`

**Comment by anoopmundathan** [comment]
Same error for me as well while running on andriod 7.0 . Here is my ```package.json```
```
"dependencies": {
    "expo": "^23.0.4",
    "react": "16.0.0",
    "react-native": "0.50.4",
    "react-native-elements": "0.19.0",
    "react-navigation": "1.0.0-beta.23",
    "react-redux": "5.0.6",
    "redux": "3.7.2",
    "redux-thunk": "2.2.0"
  }
```

```render```  function  looks as below
```
render() {
    return(
      <ScrollView
        horizontal
        pagingEnabled
        style={{flex: 1}}>
        {this.renderSlides()}
      </ScrollView>
    )
  }
```

**Comment by mjugger** [comment]
@njho 
do you mean digging into the code base for nativebase components to remove that from an effected component?

**Comment by njho** [comment]
No I mean just in the props for ScrollView

**Comment by hugtrueme** [comment]
Hey guys, you can try to follow this instruction ( https://facebook.github.io/react-native/docs/upgrading.html ) to set your project to 0.49.5, then this error will disappear.

Remember, let your version number are match each other. (https://github.com/react-community/create-react-native-app/blob/master/VERSIONS.md)

After change the version number, run 'npm install' on command line in your project directory.

**Comment by oonxt** [comment]
same error for me...
and works after set my project to 0.49.5.   
but, some other packages got error,  SafeAreaView undefined .

**Comment by andersonaddo** [comment]
@hugtrueme How are you guys downgrading? I've been trying to downgrade to 0.49.5 form 0.54.0, but when I do the whole app fails to start on my Android emulator.
My package.json:
`{
  "name": "ssnitmobileapp",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest"
  },
  "dependencies": {
    "react": "16.0.0-beta.5",
    "react-native": "0.49.5",
    "react-native-elements": "1.0.0-beta2",
    "react-native-maps": "^0.20.1",
    "react-native-vector-icons": "^4.5.0"
  },
  "devDependencies": {
    "babel-jest": "22.4.1",
    "babel-preset-react-native": "4.0.0",
    "jest": "22.4.2",
    "react-test-renderer": "16.0.0-beta.5"
  },
  "jest": {
    "preset": "react-native"
  }
}
`

**Comment by chikarao** [comment]
I'm working on a Expo-generated app and got the message message about unsupported top level event type. The the app started back up when I restarted the app on the Expo console.

**Comment by hugtrueme** [comment]
@andersonaddo, sorry for late, i posted my package.son & app.son on here.

https://github.com/facebook/react-native/issues/16522#issuecomment-368242086

**Comment by mtmcfarl** [comment]
I'm having this issue on react-native 0.52.0.  Below is my package.json.  The error occurs when I tap the back button.

```{
  "name": "App",
  "version": "0.1.0",
  "private": true,
  "devDependencies": {
    "react-native-scripts": "1.11.1",
    "jest-expo": "25.0.0",
    "react-test-renderer": "16.2.0"
  },
  "main": "./node_modules/react-native-scripts/build/bin/crna-entry.js",
  "scripts": {
    "start": "react-native-scripts start",
    "eject": "react-native-scripts eject",
    "android": "react-native-scripts android",
    "ios": "react-native-scripts ios",
    "test": "node node_modules/jest/bin/jest.js"
  },
  "jest": {
    "preset": "jest-expo"
  },
  "dependencies": {
    "expo": "^25.0.0",
    "react": "16.2.0",
    "react-native": "0.52.0"
  }
}```

**Comment by grath10** [comment]
Same issue on  react-native 0.54.0， appreciate any help or advice as to solve this headache!

**Comment by kingofglory** [comment]
After I added  a 
```
<ScrollView /> 
```
into layout , the problem solved!

**Comment by AliSarraf** [comment]
+1 I'm having the same issue - would greatly appreciate any help!

Is there any way of fixing this without having to downgrade to 0.49.5? Because I need to run at least 0.50.3 due to some other project dependencies

**Comment by hramos** [comment]
Reopening as I noticed people are claiming this is still an issue in v0.54.0 (current stable).

Let's limit new comments to those with information pertaining to v0.54.0. If you are not running v0.54.0 or newer, please abstain from commenting on the issue.

**Comment by pascalporedda** [comment]
@hramos what information do you need? As I'am experiencing this issue as well, in a really simple app. Downgrading to react-native version 0.49.5 worked fine, but in the latest 0.54.2 it was broken for me.

All I did was create an app with react-native init,  remove the scaffolding done by init and outputted some text. I'am running the app on the iPhone 6 Simulator as well as on a Google Pixel emulator, both were causing the same problem.

**Comment by hramos** [comment]
@pascalporedda if your repro steps match the original post, then I don’t think an update is necessary, other than perhaps mentioning its present in 54.2. If you have additional information on how this can be reproduced, that would help. 

Ideally, if someone digs deep into the root cause, that’s the sort of information that would be helpful in this thread. 

If your repro steps do not match the original, then I would report it as a new, distinct issue.

**Comment by AliSarraf** [comment]
After upgrading to v0.54.0, this error didn't trigger any more - thanks @hramos! :)

To give some more context, I'm setting the horizontal prop to true in the ScrollView component, this specific error was triggering on Android before when I was running v0.50.3.

**Comment by wnr** [comment]
We still encounter this issue with 0.55.4. We created a new project with `react-native init ExampleApp`, and then added two simple screens for `NavigatorIOS` to navigate between. We get the `Unsupported top level event "topScroll" dispatched` when the back button of the navigator is clicked. If we wrap the screens with `ScrollView` it works.

**Comment by eroosenmaallen** [comment]
We are seeing this error in rect-native 0.55.4.
For us, removing the key prop from child views did the trick, but it's hackish and triggers the missing key warning.

**Comment by elbsurfer** [comment]
Hello,
I use 0.56 and still see this issue on iOS. I am using NavigatorIOS - this issue is still not solved, is it?

**Comment by igrishaev** [comment]
So do I with NavigatorIOS. The red screen with the same error occurs when I go back to the previous view.

**Comment by clewiston** [comment]
I'm seeing this in react-native ^0.56.0.  Is the workaround still to wrap the NavigationIOS component in a ScrollView?

**Comment by eliakorkmaz** [comment]
version: 0.57.1, when you popBack with using NavigationIOS, this error happens.

**Comment by eliakorkmaz** [comment]
If your need are urgent and want to accomplish single page, you can solve it with using `ScrollView` and disable its scroll with `scrollEnabled='false'` for a temporary a way of running from the error. 

```xml
<ScrollView scrollEnabled='false'>
          <SafeAreaView>
            <View>
                <Text>Current Scene: {this.props.title}</Text>
                <Button
                    onPress={this._onForward}
                    title="Tap me to load the next scene"
                  />
            </View>
      </SafeAreaView>
      </ScrollView> 
```

**Comment by tapz** [comment]
Is the root cause the same for other this kind of errors? I get one with gestures with both react-navigation and my own code. Both iOS and Android. This is a show stopper, as the app does not work at all anymore.

"react": "16.8.6",
"react-native": "0.59.2",
"react-native-gesture-handler": "1.1.0",
"react-navigation": "3.5.1",

![IMG_3066](https://user-images.githubusercontent.com/11726387/55247385-2437aa80-5250-11e9-8269-fedd3a5d31ff.PNG)
